python/modem-firmware-1.3+/claim_and_provision_device.py

```diff
@@ -0,0 +1,522 @@
+        else:
+            # if a specific port not requested, filter out ports whose /dev
+            # does not end in a 1
+            if selected_port == None and not list_all and port[-1] != '1':
+                if verbose:
+                    print(send_style('Skipping: {:2}: {:20} {!r} {!r}'.
+                                      format(port_num, port, desc, hwid)))
+                continue
+
+        name = ''
+        for nm in usb_patterns:
+            if nm[0] in hwid:
+                name = nm[1]
+                break
+
+        if selected_port != None:
+            if selected_port == port:
+                return port
+        else:
+            print(send_style('{:2}: {:20} {:17}'.format(port_num, port, name)))
+            if verbose:
+                print(send_style('  {!r} {!r}'.format(desc, hwid)))
+
+            ports.append(port)
+            dev_types.append(False)
+            port_num += 1
+
+    if len(ports) == 0:
+        sys.stderr.write(error_style('No device found\n'))
+        return None
+    if len(ports) == 1:
+        return ports[0]
+    while True:
+        port = input('--- Enter port index: ')
+        try:
+            index = int(port) - 1
+            if not 0 <= index < len(ports):
+                sys.stderr.write(error_style('--- Invalid index!\n'))
+                continue
+        except ValueError:
+            pass
+        else:
+            port = ports[index]
+        return port
+
+def write_line(line, hidden = False):
+    global cmd_term_key
+    if not hidden:
+        print(send_style('-> {}'.format(line)))
+    ser.write(bytes((line + CMD_TERM_DICT[cmd_term_key]).encode('utf-8')))
+
+def wait_for_prompt(val1=b'uart:~$: ', val2=None, timeout=15, store=None):
+    global lf_done
+    found = False
+    retval = False
+    output = None
+
+    ser.flush()
+
+    while not found and timeout != 0:
+        line = ser.readline()
+
+        if line == b'\r\n':
+            # Skip the initial CRLF (see 3GPP TS 27.007 AT cmd specification)
+            continue
+
+        if line == None or len(line) == 0:
+            if timeout > 0:
+                timeout -= serial_timeout
+            continue
+
+        #sys.stdout.write('<- ' + str(line, encoding=full_encoding))
+
+        if val1 in line:
+            found = True
+            retval = True
+        elif val2 != None and val2 in line:
+            found = True
+            retval = False
+        elif store != None and (store in line or str(store) in str(line)):
+            output = line
+
+    lf_done = b'\n' in line
+    if ser:
+        ser.flush()
+    if store != None and output == None:
+        print(error_style('String {} not detected in line {}'.format(store,
+                                                                    line)))
+
+    if timeout == 0:
+        print(error_style('Serial timeout'))
+
+    return retval, output
+
+def cleanup():
+    global ser
+    ser.close()
+
+def get_attestation_token():
```

Comment:
Maybe a future improvement, but, it would be good to be able to do this without USB too.  For example, the customer provides a csv file containing one or more attestation tokens.  The script would assume the device somewhere out there in the world is on and already has the provisioning client (or a customer application that includes the provisioning library) installed as well as the proper provisioning root CA.  The script would just interact with the provisioning REST API.

Reply:
IRIS-6919